File: ZigZagConversion.java
package easy;

public class ZigZagConversion {
	public String convert(String s, int nRows) {	
		StringBuffer[] sb = new StringBuffer[nRows];
		for (int i = 0; i < sb.length; i++)	sb[i] = new StringBuffer();
		int len = s.length();
		int i = 0;
		while (i < len) {
			for (int j = 0; j < nRows && i < len; j++) {
				sb[j].append(s.charAt(i++));
			}
			for (int j = nRows - 2; j > 0 && i < len; j--) {
				sb[j].append(s.charAt(i++));
			}
		}
		for (int k = 1; k < nRows; k++)	sb[0].append(sb[k]);
		return sb[0].toString();
    }
	
	/*public static void main(String[] args) {
		ZigZagConversion zz = new ZigZagConversion();
		System.out.println(zz.convert("PAYPALISHIRING", 3));
	}*/
}

File: zigzagconversion.cpp
//
//  zigzagconversion.cpp
//  test
//

#include <stdio.h>
#include <iostream>
#include <string>
using namespace std;

class Solution {
public:
    string convert(string s, int nRows) {
        if (nRows <= 1 || s.size() <= nRows)
            return s;
        
        string result;
        
        for (int i=0; i<nRows; i++) {
            for (int j=0,index=i; index<s.size(); j++,index=(2*nRows-2)*j+i) {
                result.append(1,s[index]);
                if (i==0 || i==nRows-1) {
                    continue;
                }
                if (index+(nRows-i-1)*2<s.size()) {
                    result.append(1,s[index+(nRows-i-1)*2]);
                }
            }
        }
        return result;
    }
};

/*
int main(int argc, const char * argv[]) {
    // insert code here...
    string s = "PAYPALISHIRING";
    Solution* sol = new Solution();
    string rst = sol->convert(s, 3);
    std::cout << rst;
    return 0;
}*/
